drivers: tee: optee: Introducing OPTEE_RPC_CMD function handles

oleksiimoisieiev opened the pull request:
This includes the implementation of the basic RPC_CMD which are received from OP-TEE TA service. Memory allocation requests, gettime and suspend were implemented in this commit.

oleksiimoisieiev committed:
drivers: tee: optee: Moving optee_rpc_cmd header

Moving header with OPTEE_RPC* defines needed to handle requests from
OP-TEE.

Signed-off-by: Oleksii Moisieiev <oleksii_moisieiev@epam.com>
Acked-by: Mykola Kvach <mykola_kvach@epam.com>
Reviewed-by: Volodymyr Babchuk <volodymyr_babchuk@epam.com>

oleksiimoisieiev committed:
drivers: tee: optee: Introducing OPTEE_RPC_CMD function handles

This includes the implementation of the basic RPC_CMD which are
received from OP-TEE TA service.
Memory allocation requests, gettime and suspend were implemented in
this commit.

Signed-off-by: Oleksii Moisieiev <oleksii_moisieiev@epam.com>
Acked-by: Mykola Kvach <mykola_kvach@epam.com>
Reviewed-by: Volodymyr Babchuk <volodymyr_babchuk@epam.com>

oleksiimoisieiev committed:
tests: drivers: optee: Unit tests for OPTEE_RPC_CMD_* handlers

This tests the behaviour of the rpc handlers, based on the Zephyr unit
tests architecture.

Signed-off-by: Oleksii Moisieiev <oleksii_moisieiev@epam.com>
Reviewed-by: Volodymyr Babchuk <volodymyr_babchuk@epam.com>

lorc commented on drivers/tee/optee/optee.c:
```diff
@@ -226,12 +392,21 @@ static int optee_call(const struct device *dev, struct optee_msg_arg *arg)
+		 * Clean allocated pages if needed. Some function calls requires pages
+		 * allocation which should be freed after processing new request
+		 */
+		if (pages) {
```
How you know that OP-TEE finished reading from list already?

oleksiimoisieiev replied:
good point. I'll take a look on it. Thanks

lorc replied:
It is safe to free this list only after you get another SHM op (e,g. another alloc or free)

oleksiimoisieiev replied:
Thanks. fixed

oleksiimoisieiev committed:
drivers: tee: optee: Introducing OPTEE_RPC_NOTIFICATION handlers

OPTEE_RPC_NOTIFICATION works as the synchronization for the Secure
World. It allows OP-TEE to request WAIT command, which makes Normal
World to sleep until different thread request for OP-TEE will cause
SEND commend to be sent which will unlock waiting thread.

Signed-off-by: Oleksii Moisieiev <oleksii_moisieiev@epam.com>
Acked-by: Mykola Kvach <mykola_kvach@epam.com>
Reviewed-by: Volodymyr Babchuk <volodymyr_babchuk@epam.com>

lorc commented on drivers/tee/optee/optee.c:
```diff
@@ -177,9 +178,172 @@ static void u64_to_regs(uint64_t val, uint32_t *reg0, uint32_t *reg1)
+
+	/*
+	 * TODO: OPTEE_RPC_SHM_TYPE_KERNEL and OPTEE_SHM_TYPE_APPL should be handled
+	 * Currectly TYPE_KERNEL was implemented.
```
"Currently"

```
(dolist (hook '(c++-mode-hook c-mode-hook))
  (add-hook hook (lambda () (flyspell-prog-mode))))
``` 
;)

oleksiimoisieiev replied:
Thanks :). Hard way to become emacs guru,

lorc commented on drivers/tee/optee/optee.c:
```diff
@@ -177,9 +178,172 @@ static void u64_to_regs(uint64_t val, uint32_t *reg0, uint32_t *reg1)
+static void handle_cmd_notify(const struct device *dev, struct optee_msg_arg *arg)
+{
+	/* TODO: Implement wait for completion based on zephyr workq */
+	arg->ret = TEEC_ERROR_NOT_IMPLEMENTED;
```
You can return `TEEC_SUCCESS` here. But sleep for a 10 milliseconds before

OP-TEE has protection against spurious wakeups, so mutexes will  work... Sort of.

oleksiimoisieiev replied:
+

lorc replied:
As I said, you can return `TEEC_SUCCESS` in this case.

oleksiimoisieiev replied:
fixed

sa-kib commented on drivers/tee/optee/optee_rpc_cmd.h:
this header is incomplete, e.g. it misses OPTEE_RPC_FS_* calls, needed for supplicant FS_CMD support

oleksiimoisieiev replied:
They are in optee_msg_supplicant.h and I couldn't find it in linux driver. Are you sure it should be in Zephyr and not in supplicant?

oleksiimoisieiev committed:
tests: drivers: optee: Unit tests for OPTEE_CMD_NOTIFICATION

Tests lock/unlock mechanism for OP-TEE. Process lock/unlock from
different threads and tests send bit when SEND was received before
WAIT command.

Signed-off-by: Oleksii Moisieiev <oleksii_moisieiev@epam.com>
Reviewed-by: Volodymyr Babchuk <volodymyr_babchuk@epam.com>

xakep-amatop commented on tests/kernel/context/src/main.c:
```diff
@@ -114,7 +114,7 @@ static ISR_INFO isr_info;
  * @brief Handler to perform various actions from within an ISR context
  *
- * This routine is the ISR handler for isr_handler_trigger(). It performs
+] * This routine is the ISR handler for isr_handler_trigger(). It performs
```
looks like a typo

oleksiimoisieiev replied:
Fixed. thanks

xakep-amatop commented on tests/net/traffic_class/src/main.c:
```diff
@@ -11,7 +11,7 @@
 LOG_MODULE_REGISTER(net_test, NET_LOG_LEVEL);
 
-#include <zephyr/types.h>
+#include <zephyr/types.h>z
```
look like this is an extra typo
and this? I suppose,  breaks compilation

oleksiimoisieiev replied:
Fixed. thanks

xakep-amatop commented on drivers/tee/optee/optee.c:
```diff
@@ -27,6 +30,14 @@ LOG_MODULE_REGISTER(optee);
+ * was received before WAIT command from OP-TEE. In this case WAIT will not create
+ * locks.
+ */
+SYS_BITARRAY_DEFINE_STATIC(notif_bitmap, CONFIG_OPTEE_MAX_NOTIF);
```
wouldn't it be better to put it to  `optee_driver_data` if you will have a few driver instances?
according to the GP TEE documentation it is possible to have a few TEEs, e.g. two optee OSs running in one board, but on their schemes there is only one driver instance per all TEEs:
![Screenshot from 2023-06-29 13-19-26](https://github.com/oleksiimoisieiev/zephyr/assets/27217224/8e21a00e-cbb1-4310-8c84-a8068ec27ef9)



as least I think it would be nice if you could save modularity of drivers if we have a few driver instances in a future

oleksiimoisieiev replied:
Mentioned this in https://github.com/xen-troops/zephyr/pull/55 as TODO

xakep-amatop commented on drivers/tee/optee/optee.c:
```diff
@@ -177,9 +197,305 @@ static void u64_to_regs(uint64_t val, uint32_t *reg0, uint32_t *reg1)
+		handle_cmd_get_time(dev, arg);
+		break;
+	case OPTEE_RPC_CMD_NOTIFICATION:
+		handle_cmd_notify(dev, arg);
```
I hope this and below case isn't possible for atomic context

lorc replied:
Atomic context where exactly?

xakep-amatop replied:
in case if somebody invokes optee call under spin lock or from IRQ handler and the OS sends RPC
and request require some sleep

lorc replied:
It is programmer's error to invoke TEE from an atomic context. Exactly because TEE calls can be blocking.

xakep-amatop commented on drivers/tee/optee/optee.c:
```diff
@@ -177,9 +197,305 @@ static void u64_to_regs(uint64_t val, uint32_t *reg0, uint32_t *reg1)
+		goto out;
+	}
+
+	*pages = pl;
```
who will free pl and when ?
pls, show me call chain

xakep-amatop reviewed (changes requested):
review isn't finished yet

xakep-amatop commented:
> It is safe to free this list only after you get another SHM op (e,g. another alloc or free)
@oleksiimoisieiev pls, revert changes that have been made for this comment

I believe that you should revert the next changes and delete all the newly added code related to the deletion:
![Screenshot from 2023-06-29 16-08-45](https://github.com/oleksiimoisieiev/zephyr/assets/27217224/73879b86-73a6-4732-b3d4-3f4f80628373)


First of all, there are a few issues with the current implementation. One problem is that the memory for the page list isn't freed in the following cases:
* when you allocate mem using RPC and it was the last RPC in optee_call, you lose the pointer because you leave `optee_call` without storing it anywhere;
* and also lose the pointer to the page list if, after an RPC for memory allocation, you receive a non-memory management RPC and leave `optee_call`.

Regarding the safety concern, it should be noted that OP-TEE OS copies all pages to its own OS structures. You can observe this in the following call chain:
`thread_rpc_alloc` --> `get_rpc_alloc_res` --> `rpc_shm_mobj_alloc` --> `mobj_mapped_shm_alloc` --> `mobj_reg_shm_alloc`:
```
memcpy(mobj_reg_shm->pages, pages, sizeof(*pages) * num_pages);
```
Furthermore, both U-Boot and Linux kernel drivers free the page list inside an analogous function to our Zephyr function `optee_call`.

Note: the page list memory is used solely for communication purposes, and we do not need to retain it until the free command.

lorc commented:
> Regarding the safety concern, it should be noted that OP-TEE OS copies all pages to its own OS structures. You can observe this in the following call chain:

Yes, but you must understand that this is asynchronous process. OP-TEE may return with another RPC due to an interrupt before it have chance to copy from the list. This is why Linux drivers calls `free_pages_list` during another `OPTEE_MSG_RPC_CMD_SHM_ALLOC`:

```
	case OPTEE_MSG_RPC_CMD_SHM_ALLOC:
		free_pages_list(call_ctx);
		handle_rpc_func_cmd_shm_alloc(ctx, arg, call_ctx);
		break;
```

Because only in this case we can be sure that it already finished handling the previous `alloc` call and read from the list.

And of course the list needs to be freed if we leave `optee_call`.

xakep-amatop commented:
> > Regarding the safety concern, it should be noted that OP-TEE OS copies all pages to its own OS structures. You can observe this in the following call chain:
> 
> Yes, but you must understand that this is asynchronous process. OP-TEE may return with another RPC due to an interrupt before it have chance to copy from the list. This is why Linux drivers calls `free_pages_list` during another `OPTEE_MSG_RPC_CMD_SHM_ALLOC`:
> 
> ```
> 	case OPTEE_MSG_RPC_CMD_SHM_ALLOC:
> 		free_pages_list(call_ctx);
> 		handle_rpc_func_cmd_shm_alloc(ctx, arg, call_ctx);
> 		break;
> ```
> 
> Because only in this case we can be sure that it already finished handling the previous `alloc` call and read from the list.
> 
> And of course the list needs to be freed if we leave `optee_call`.

I see the issue regarding the safety concern, thanks.

In the situation you described, there is a potential problem with memory management and synchronization between OPTEE OS and the Linux kernel driver. Here's a revised breakdown of the steps with the potential problems highlighted:

    You invoke a command to OPTEE OS.
    OPTEE OS handles the command and requests an RPC (Remote Procedure Call) for memory allocation and registration.
    You return a buffer to OPTEE OS for registration.
    Right before the copying process (call to thread_rpc_alloc), an interrupt request (IRQ) is received.
    The IRQ triggers an allocation memory RPC in the Linux kernel.
    The Linux kernel driver frees the memory context for the previous unfinished command.
    OPTEE OS copies the buffer, but since the memory context has been freed, it may contain garbage or inconsistent data.

so, is it situation possible or not ?

lorc commented:
> The IRQ triggers an allocation memory RPC in the Linux kernel.

This is not possible. IRQ always comes from a device and is handled by a device driver. There is no chance it can cause RPC call inside the same STD call context.

xakep-amatop commented on drivers/tee/optee/optee.c:
```diff
@@ -177,9 +197,304 @@ static void u64_to_regs(uint64_t val, uint32_t *reg0, uint32_t *reg1)
+		return -EINVAL;
+	}
+
+	sp_key = k_spin_lock(&data->notif_lock);
+
+	SYS_DLIST_FOR_EACH_CONTAINER(&data->notif, iter, node) {
+		if (iter->key == key) {
+			k_sem_give(&iter->wait);
+			goto out;
+		}
+	}
+
+	/* If nobody is waiting for key - set bit in the bitmap */
+	sys_bitarray_set_bit(&notif_bitmap, key);
+out:
+	k_spin_unlock(&data->notif_lock, sp_key);
```
```suggestion
	sp_key = k_spin_lock(&data->notif_lock);
	if (!key_is_pending(data, key)) {
		/* If nobody is waiting for key - set bit in the bitmap */
		sys_bitarray_set_bit(&notif_bitmap, key);
	}
	k_spin_unlock(&data->notif_lock, sp_key);
```

oleksiimoisieiev replied:
+1 thanks. I'll update.

xakep-amatop commented on drivers/tee/optee/optee.c:
```diff
@@ -177,9 +197,304 @@ static void u64_to_regs(uint64_t val, uint32_t *reg0, uint32_t *reg1)
+	}
+
+	/*
+	 * TODO: OPTEE_RPC_SHM_TYPE_KERNEL and OPTEE_SHM_TYPE_APPL should be handled
```
as far as I see you have already handled `OPTEE_RPC_SHM_TYPE_KERNEL` here and in `handle_cmd_free` function

oleksiimoisieiev replied:
I left this comment to explicitly note that this expects different behavior for SHM types. I'll rephrase.

oleksiimoisieiev replied:
There is another line it this todo. Saying that type_kernel is implemented

xakep-amatop replied:
better to delete it from todo and delete below line
just leave the comment that it is needed to handle `OPTEE_SHM_TYPE_APPL` only

oleksiimoisieiev replied:
fixed

xakep-amatop commented on drivers/tee/optee/optee.c:
```diff
@@ -177,9 +197,304 @@ static void u64_to_regs(uint64_t val, uint32_t *reg0, uint32_t *reg1)
+	}
+
+	/*
+	 * TODO: OPTEE_RPC_SHM_TYPE_KERNEL and OPTEE_SHM_TYPE_APPL should be handled
```
see above comment

oleksiimoisieiev replied:
see above

lorc commented on drivers/tee/optee/optee.c:
```diff
@@ -177,9 +178,191 @@ static void u64_to_regs(uint64_t val, uint32_t *reg0, uint32_t *reg1)
 }
 
-static void handle_rpc_call(const struct device *dev, struct optee_rpc_param *param)
+static inline bool check_param_input(struct optee_msg_arg *arg)
```
So, it returns `false` if parameters are correct? This is counter-intuitive

oleksiimoisieiev replied:
fixed

xakep-amatop commented:
so, some of comments aren't addressed and you have already a new one pull request https://github.com/xen-troops/zephyr/pull/60

oleksiimoisieiev commented:
> so, some of comments aren't addressed and you have already a new one pull request [xen-troops#60](https://github.com/xen-troops/zephyr/pull/60)

I did. to the wip branch

oleksiimoisieiev commented:
Updated PR. Considering that optee_upstr was already merged to zephyr-v3.3.0-xt branch, this PR should be recreated to xen-troops/zephyr-v3.3.0-xt. But I'm leaving it as is to save comments history. Once I get Acks and RB, I'll close this mr and push already reviewed code to xen-troops

lorc commented on drivers/tee/optee/optee.c:
```diff
@@ -177,9 +178,189 @@ static void u64_to_regs(uint64_t val, uint32_t *reg0, uint32_t *reg1)
-static void handle_rpc_call(const struct device *dev, struct optee_rpc_param *param)
+static inline bool check_param_input(struct optee_msg_arg *arg)
+{
+	return !(arg->num_params != 1 ||
+		 arg->params[0].attr != OPTEE_MSG_ATTR_TYPE_VALUE_INPUT);
```
Oh well... Any particular reason why you can't rewrite it as 
```
return arg->num_params == 1 &&
		 arg->params[0].attr == OPTEE_MSG_ATTR_TYPE_VALUE_INPUT
```
? 

It is really hard to read and do all logical conversions in my head.

This is the security-related code in a safety-oriented OS. Code must be as simple and as dumb as possible

oleksiimoisieiev replied:
fixed

lorc commented on drivers/tee/optee/optee.c:
```diff
@@ -177,9 +178,189 @@ static void u64_to_regs(uint64_t val, uint32_t *reg0, uint32_t *reg1)
+		return TEEC_ERROR_NOT_IMPLEMENTED;
+	default:
+		/* TODO: supplicant commands are still not implemented */
+		/* handle_rpc_supp_cmd(ctx, optee, arg); */
```
No commented out code, please.  Have you ran checkpatch, BTW?

xakep-amatop replied:
```
Rule 2.2 | Required | There shall be no dead code
```

oleksiimoisieiev replied:
fixed

lorc commented on drivers/tee/optee/optee.c:
```diff
@@ -177,9 +178,189 @@ static void u64_to_regs(uint64_t val, uint32_t *reg0, uint32_t *reg1)
+
+static void handle_cmd_wait(const struct device *dev, struct optee_msg_arg *arg)
+{
+	if (check_param_input(arg)) {
```
Missed inversion here

oleksiimoisieiev replied:
fixed

lorc commented on drivers/tee/optee/optee.c:
```diff
@@ -286,19 +305,125 @@ static void handle_cmd_get_time(const struct device *dev, struct optee_msg_arg *
 static void handle_cmd_wait(const struct device *dev, struct optee_msg_arg *arg)
 {
-	if (check_param_input(arg)) {
+	if (!check_param_input(arg)) {
```
This is the wrong commit for this change

oleksiimoisieiev replied:
fixed

xakep-amatop reviewed (approved):
`Acked-by: Mykola Kvach <mykola_kvach@epam.com>`
only the next commits:
94ac9994c16b35ad474292102b99bb4f442ce7d4
8fdd176099dab12e5f9b0dd7270d6b78230933b1
5a569e5a08f1c5f7efcd7f0fdaa58ce151d47479

lorc reviewed (approved):
`Reviewed-by: Volodymyr Babchuk <volodymyr_babchuk@epam.com>`

oleksiimoisieiev commented:
Thank you for the review. I'm closing this PR to recreate it in xen-troops repo
